arrow/src/main/java/org/apache/iceberg/arrow/DictEncodedArrowConverter.java

```diff
@@ -0,0 +1,224 @@
+import java.util.function.IntConsumer;
+import org.apache.arrow.vector.BaseFixedWidthVector;
+import org.apache.arrow.vector.BaseVariableWidthVector;
+import org.apache.arrow.vector.BigIntVector;
+import org.apache.arrow.vector.DecimalVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.Float4Vector;
+import org.apache.arrow.vector.Float8Vector;
+import org.apache.arrow.vector.TimeMicroVector;
+import org.apache.arrow.vector.TimeStampMicroTZVector;
+import org.apache.arrow.vector.TimeStampMicroVector;
+import org.apache.arrow.vector.TimeStampVector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.arrow.vector.VarCharVector;
+import org.apache.iceberg.arrow.vectorized.ArrowVectorAccessor;
+import org.apache.iceberg.arrow.vectorized.VectorHolder;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/** This converts dictionary encoded arrow vectors to a correctly typed arrow vector. */
+public class DictEncodedArrowConverter {
+
+  private DictEncodedArrowConverter() {}
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static FieldVector toArrowVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Preconditions.checkArgument(null != vectorHolder, "VectorHolder cannot be null");
+    Preconditions.checkArgument(null != accessor, "ArrowVectorAccessor cannot be null");
+    if (vectorHolder.isDictionaryEncoded()) {
+      if (Type.TypeID.DECIMAL.equals(vectorHolder.icebergType().typeId())) {
+        return toDecimalVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIMESTAMP.equals(vectorHolder.icebergType().typeId())) {
+        return toTimestampVector(vectorHolder, accessor);
+      } else if (Type.TypeID.LONG.equals(vectorHolder.icebergType().typeId())) {
+        return toBigIntVector(vectorHolder, accessor);
+      } else if (Type.TypeID.FLOAT.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat4Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.DOUBLE.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat8Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.STRING.equals(vectorHolder.icebergType().typeId())) {
+        return toVarCharVector(vectorHolder, accessor);
+      } else if (Type.TypeID.BINARY.equals(vectorHolder.icebergType().typeId())) {
+        return toVarBinaryVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIME.equals(vectorHolder.icebergType().typeId())) {
+        return toTimeMicroVector(vectorHolder, accessor);
+      }
+
+      throw new IllegalArgumentException(
+          String.format(
+              "Cannot convert dict encoded field '%s' of type '%s' to Arrow "
+                  + "vector as it is currently not supported",
+              vectorHolder.icebergField().name(), vectorHolder.icebergType().typeId()));
```

Comment:
I think this case would get hit even in the case the vector is not dictionary encoded and we'd be surfacing a misleading error message, would it make sense to throw a different exception to indicate that this converter should not be used for non dictionary encoded vectors?

Reply:
Missed that check in that case we should be fine.